Lecture on automation (GitHub Actions, GitLab CI) (#78)
Simulation-Software-Engineering/Lecture-Material: 05_testing_and_ci/github_actions_gitlab_ci_demo.md

# Demo: Automation with GitHub Actions and GitLab CI

## Automation

> - Nothing to show

Reviewer: This seems unnecessary.

## Preliminary step --- Building a tiny Docker container

> Goal: Configure a Docker container useful for our tests. This should show
> 
> - why we did introduce Docker before.
> - a common use case of Docker.
> - Show how to make even tinier containers as before. We use Alpine Linux here.

Reviewer: ```suggestion
Goal: Configure a Docker container useful for our tests. This should show ...

- why we did introduce Docker before.
- a common use case of Docker.
- how to make even tinier containers as we did previously. We use Alpine Linux here.
```

Steps:

> - Skip this if more time is needed

Reviewer: Do you know this at this point already? :grimacing:

  ```

- Now the image can be used as `ajaust/automation-lecture` or `ajaust/automation-lecture:latest`. I did not version the uploaded container.

## GitHub Actions

### 1. Setting up a test job

- Go to [repository](https://github.com/Simulation-Software-Engineering/automation-lecture) an check out tag `initial-state`.
- Create a new branch `lecture-branch`
- Set up workflow file

  ```bash
  mkdir -p .github/workflows
  cd .github/workflows
  touch testing.yml
  ```

- Run `black` shortly locally and explain what it does.

  ```bash
  black --check .
  ```

    - It is a style checker that checks whether the Python files are formatted according to a certain standard.
    - `black --check .` will check files without modifying them.
    - `black` uses a lot of smileys

- Edit `testing.yml` to have following content

  ```yaml
  name: Testing workflow

  on: [push, pull_request]

  jobs:
    test:
      runs-on: ubuntu-latest
      steps:
        - uses: actions/checkout@v2
        - uses: actions/setup-python@v2
          with:
            python-version: '3.8.10'
        - name: "Install pytest dependency"
          run: pip install --user  pytest
        - name: "Run unittest"

>           run: python -m unittest

Reviewer: You install pytest, but run unitest. That might confuse some. See comment above: I would suggest to simply drop one of them.

Author: It is gone.

  ```

- `runs-on` does **not** refer to a Docker container, but to runner tag.
- After the push inspect "Action" panel on GitHub repository
    - GitHub will schedule a run (yellow dot)
    - Hooray. We have set up our first action.

> - Edit `test_operations.py`, break a test, commit it and push it to the repository. Check the failing test.

Reviewer: Actually, it could be nicer from a reality viewpoint to break the implementation, not the test. Add to do it in a PR? And maybe to set the options in the repo that you cannot merge then?

### 2. Extend Action to have several dependent jobs

> - Go to tag `github-action-added` or go to `main` branch of the repository. The workflow should have the following content:

Reviewer: ```suggestion
- Go to tag `github-action-added` of the repository. The workflow should have the following content:
```
To keep it simple and maybe we will add more in the future.

  ```yaml
  name: Testing workflow

  on: [push, pull_request]

  jobs:
    style:
      runs-on: ubuntu-latest
      steps:
        - uses: actions/checkout@v2
        - uses: actions/setup-python@v2
          with:
            python-version: '3.8.10'
        - name: "Install style checker"
          run: pip install --user black
        - name: "Run style check"
          run: black --check .
    build:
      needs: style
      runs-on: ubuntu-latest
      env:
        PROJECT_NAME: "Automation Lecture"
      steps:
        - name: "Run build phase"
          run: echo "Building project $PROJECT_NAME"
    test:
      needs: build
      runs-on: ubuntu-latest
      steps:
        - uses: actions/checkout@v2
        - uses: actions/setup-python@v2
          with:
            python-version: '3.8.10'
        - name: "Install pytest dependency"
          run: pip install --user  pytest
        - name: "Run unittest"
          run: python -m unittest
  ```

- We need to run `actions/checkout@v2` in each job
    - We could share the repository between jobs via artifacts, but that is uncommon.
- We need to run `actions/setup-python@v2` since jobs do not share the environment

> - We specify dependencies `need` such that the steps run after each other

Reviewer: ```suggestion
- We specify dependencies by `needs` such that the steps run after each other.
```
